ql/src/java/org/apache/hadoop/hive/ql/PathCleaner.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class is used to asynchronously remove directories after query execution
+ */
+public class PathCleaner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PathCleaner.class.getName());
+  private static final AsyncDeleteAction END_OF_PROCESS = new AsyncDeleteAction(null, null);
+
+  private final BlockingDeque<AsyncDeleteAction> deleteActions = new LinkedBlockingDeque<>();
+  private final AtomicBoolean isShutdown = new AtomicBoolean();
```

Comment:
note: all usages are negated for this boolean - instead of using negative logic (shutdown) ; using a positive name like "run" might make things easier to read/follow